packages/flutter_tools/lib/src/android/android_device.dart

```diff
@@ -699,6 +703,34 @@ class AndroidDevice extends Device {
     }
   }
 
+  /// Query for the launched application's PID.
+  ///
+  /// This may need to run several times, since the application does not immediately
+  /// appear in the PID query after launch. If this continually fails, we can fall
+  /// back to using the VM Service PID query.
+  ///
+  /// Returns `null` if the query fails.
+  Future<int> _queryForPid(AndroidApk package) async {
+    while (true) {
```

Comment:
I'm not sure, but I should probably limit this to a low number of attempts

Reply:
Yeah, it may crash before you get a chance to query it, then this will never exit successfully.

If you add `-W` to `adb shell start` to wait for the process to launch does `adb shell pidof` always return the pid?

Reply:
If so, it would be nice to be able to confidently `LaunchResult.failed()` if this is null instead of doing any vm service fallbacks.

Reply:
agreed - I need to test this with some crashing applications to see what the behavior is like

Reply:
A native crash app integration test would be a good idea, like https://github.com/flutter/flutter/blob/3ddd77deb662054dc2c4096b4f6d592ee18b5235/packages/flutter_tools/test/integration.shard/flutter_run_with_error_test.dart but from the native code instead of from Flutter.  I swear I wrote something like that at some point but it must have been a hallucination.